### PR TITLE
Features/stub method interfaces 177542925

### PR DESCRIPTION
**Pull request by rosalinekarr**
Problem
=======
We don't have a large architectural plan for modules and specific methods in the SDK, and that lack makes it difficult to start work on implementations when the interfaces they need to interact with are unclear.
[#177542925](https://www.pivotaltracker.com/story/show/177542925)

Solution
========
Add interfaces, stubs and documentation for as many modules and methods as we can think of for the v1 release. These interfaces and methods will likely change over the course of implementation, but their basic outline at least gives us a place to start and general concept of what other method calls will be available to us in implementation.

with @wilwade

### CODE REVIEW

File: src/utilities/hash.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export type KeccakHash = string;
```
Comment (shannonwells) [inline src/utilities/hash.ts:1]: This type is exported in its own file already

File: src/types/hash.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+/**
+ * KeccakHash is a 256 bit hash string encoded as hex with a '0x' at the start
+ *
+ * @type KeccakHash
+ */
+export type KeccakHash = string;
```
Comment (shannonwells) [inline src/types/hash.ts:6]: If I understand how this is supposed to work correctly, I think we should follow putting this into type declarations i.e put them in a .d.ts file and do:
```
declare type KeccakHash = string
```

File: src/activityPub/activityPub.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import { keccak256 } from "js-sha3";
+import { KeccakHash } from "../utilities/hash";
+import { NotImplementedError } from "../utilities/errors";
+
+const sortJSON = (obj: Record<string, unknown>): Record<string, unknown> => {
```
Comment (enddynayn) [inline src/activityPub/activityPub.ts:5]: can we move this into utility file?

File: src/config/config.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// import Web3 from "web3";
+// import MemoryStore from "../storage/memoryStorage";
```
Comment (enddynayn) [inline src/config/config.ts:2]: maybe we can use node-convict for configuration?

File: src/utilities/hash.ts (hunk 1 above)
Comment (rosalinekarr) [reply src/utilities/hash.ts]: Fixed

File: src/types/hash.ts (hunk 2 above)
Comment (rosalinekarr) [reply src/types/hash.ts]: Fixed

File: src/activityPub/activityPub.ts (hunk 3 above)
Comment (rosalinekarr) [reply src/activityPub/activityPub.ts]: Yep, fixed

File: src/config/config.ts (hunk 4 above)
Comment (rosalinekarr) [reply src/config/config.ts:2]: sure, I think it's worth a research spike. would you mind adding a chore to the backlog?

Comment (wilwade) [reply src/config/config.ts:2]: I took a quick look and I liked what I saw. Added it to the backlog to discuss https://www.pivotaltracker.com/story/show/177682091

**Review by enddynayn — APPROVED** [review]

File: src/config/config.ts (hunk 4 above)
Comment (shannonwells) [reply src/config/config.ts:2]: Same here, looks good at first glance

**Review by shannonwells — APPROVED** [review]
Just a couple of comments - thanks for doing this.  I agree I'm sure this isn't the final form so I think this is a great start.

File: src/batch/batch.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * @returns         A batch file object
+ */
+export const createFile = (_events: MessageType[]): BatchFileObject => {
+  throw NotImplementedError;
+};
+
+/**
+ * openFile() takes a Batch file object . This method is not yet implemented.
+ *
+ * @params  file  An array of DSNPEvent to include in the batch file
```
Comment (aneyzberg) [inline src/batch/batch.ts:20]: is this description correct? It seems like a copy and past error from above

File: src/batch/batch.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * @returns       A batch file object
```
Comment (aneyzberg) [inline src/batch/batch.ts:21]: It seems like the return value might be incorrect here based on the message signature below?

**Review by aneyzberg — APPROVED** [review]
lgtm just a couple small comments

File: src/contracts/announcement.ts [hunk 7]
```diff
@@ -4,8 +4,8 @@ import Web3 from "web3";
 import { Contract } from "web3-eth-contract";
 import { AbiItem } from "web3-utils";
 
-import { KeccakHash } from "../types/hash";
-import { hashPrefix } from "../utilities/hashPrefix";
+import { HexString } from "../types/Strings";
```
[the remark this answers is not in the document]
Comment (rosalinekarr) [reply src/contracts/announcement.ts:7]: Not sure, but I think we can save making a decision in a future PR.

File: src/batch/batch.ts (hunk 5 above)
Comment (rosalinekarr) [reply src/batch/batch.ts]: fixed

File: src/batch/batch.ts (hunk 6 above)
Comment (rosalinekarr) [reply src/batch/batch.ts]: fixed

**Pull request merged**